apps/worker/src/publisher/ipfs.publisher.ts

```diff
@@ -0,0 +1,86 @@
+import { ConfigService } from '../../../api/src/config/config.service';
+import { IPublisherJob } from '../interfaces/publisher-job.interface';
+import { createKeys } from '../blockchain/create-keys';
+
+@Injectable()
+export class IPFSPublisher {
+  private logger: Logger;
+
+  constructor(private configService: ConfigService, private blockchainService: BlockchainService, private eventEmitter: EventEmitter2) {
+    this.logger = new Logger(IPFSPublisher.name);
+  }
+
+  public async publish(messages: IPublisherJob[]): Promise<{ [key: string]: bigint }> {
+    const providerKeys = createKeys(this.configService.getProviderAccountSeedPhrase());
+
+    let batch: SubmittableExtrinsic<'rxjs', ISubmittableResult>[] = [];
+    const batches: SubmittableExtrinsic<'rxjs', ISubmittableResult>[][] = [];
+    const allowedBatchLen = await this.blockchainService.capacityBatchLimit();
+    messages.forEach((message) => {
+      batch.push(this.blockchainService.createExtrinsicCall({ pallet: 'messages', extrinsic: 'addIpfsMessage' }, message.schemaId, message.data.cid, message.data.payloadLength));
+
+      if (batch.length === allowedBatchLen) {
+        batches.push(batch);
+        batch = [];
+      }
+    });
+
+    if (batch.length > 0) {
+      batches.push(batch);
+    }
+    return this.sendAndProcessChainEvents(providerKeys, batches);
+  }
+
+  async sendAndProcessChainEvents(providerKeys: KeyringPair, batchesMap: SubmittableExtrinsic<'rxjs', ISubmittableResult>[][]): Promise<{ [key: string]: bigint }> {
+    try {
+      // iterate over batches and send them to the chain returning the capacity withdrawn
+      const batchPromises: Promise<{ [key: string]: bigint }>[] = [];
+
+      batchesMap.forEach(async (batch) => {
+        batchPromises.push(this.processSingleBatch(providerKeys, batch));
+      });
+
+      const totalCapUsedPerEpoch = await Promise.all(batchPromises);
+      const totalCapacityUsed = totalCapUsedPerEpoch.reduce((acc, curr) => {
+        const epoch = Object.keys(curr)[0];
+        if (acc[epoch]) {
+          acc[epoch] += curr[epoch];
+        }
+        acc[epoch] = curr[epoch];
+        return acc;
+      }, {} as { [key: string]: bigint });
+
+      this.logger.debug(`Total capacity used: ${JSON.stringify(totalCapacityUsed)}`);
+      return totalCapacityUsed;
+    } catch (e) {
+      this.logger.error(`Error processing batches: ${e}`);
+      throw e;
+    }
+  }
+
+  async processSingleBatch(providerKeys: KeyringPair, batch: SubmittableExtrinsic<'rxjs', ISubmittableResult>[]): Promise<{ [key: string]: bigint }> {
+    this.logger.debug(`Submitting batch of size ${batch.length}`);
+    try {
+      const currrentEpoch = await this.blockchainService.getCurrentCapacityEpoch();
+      const [event, eventMap] = await this.blockchainService
+        .createExtrinsic({ pallet: 'frequencyTxPayment', extrinsic: 'payWithCapacityBatchAll' }, { eventPallet: 'utility', event: 'BatchCompleted' }, providerKeys, batch)
```

Comment:
I'm not totally sure about benefits of using batch in here since we are already batching in prior steps. I think using `payWithCapacityBatchAll` might complicate the exact events and errors that we might get from chain. I think since the batching is already done I wouldn't want to change it but if in future it complicates error handling we might need to revisit it.

Reply:
This is a capacity batching, basically instead of sending one IPFS message to frequency we send as many as capacity accepts in batch (10 for now) than sending one tx at a time

Reply:
Also currently we are only sending one message at a time, I kept it like reconnection service, in case we would want to batch more than one

Reply:
I see, I think what I might have been missing is understanding the benefits of capacity batching for our scenario. I know we can use it for transactional purposes if we want a few calls apply at the same time but that is not that relevant for us since each message is treated independently. Are there any other benefits to capacity batching besides transaction?